src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java

```diff
@@ -162,6 +185,39 @@ public void arcadeDrive(double speed, double rotation, boolean squareInputs) {
     m_drive.arcadeDrive(speed, rotation, squareInputs);
   }
 
+  /**
+   * Drives the left and right wheels at the specified velocities using closed loop velocity control
+   * on the Spark Maxes.
+   *
+   * @param leftVelocity The velocity of the left wheels, in meters per second.
+   * @param rightVelocity The velocity of the right wheels, in meters per second.
+   */
+  public void tankDriveVelocities(double leftVelocity, double rightVelocity) {
+    // Use a linear model of the drivetrain to calculate feedforward voltages
+    // TODO: Handle cases where the feedforward voltage is over 12V
+    Vector<N2> velocities = VecBuilder.fill(leftVelocity, rightVelocity);
+    Matrix<N2, N1> feedforwards = m_feedforward.calculate(velocities);
+
+    // TODO: Test if the static component makes other paths more accurate
+    double leftVolts = feedforwards.get(0, 0) + Math.signum(leftVelocity) * kSVolts;
+    double rightVolts = feedforwards.get(1, 0) + Math.signum(rightVelocity) * kSVolts;
+
+    // Command the motors at the above feedforward voltages, using PID to correct for error
+    m_leftController.setReference(leftVelocity, CANSparkMax.ControlType.kVelocity, 0, leftVolts);
```

Comment:
Can you explain this to me a little more? I don't get why we need to send both velocity and voltage down to the controller.

Reply:
Yeah sure! The voltage is just the feedforward voltage supplied to the motor controller, which allows us to account for acceleration and static friction in the feedforward calculations (as well as use `LinearPlantInversionFeedforward`). The velocity is fed in for (velocity) PID control to be done on the Spark Maxes themselves, since as we've discussed they refresh at a higher rate than the main robot loop.

There is a way to get a higher refresh rate on the rio itself, but it takes a lot of boilerplate to get working in command-based.

Let me know if that all made sense :)